src/main/java/appeng/fluids/parts/PartFluidLevelEmitter.java

```diff
@@ -0,0 +1,325 @@
+	public static final PartModel MODEL_ON_HAS_CHANNEL = new PartModel( MODEL_BASE_ON, MODEL_STATUS_HAS_CHANNEL );
+
+	private static final int FLAG_ON = 4;
+
+	private boolean prevState = false;
+	private long lastReportedValue = 0;
+	private long reportingValue = 0;
+	private IStackWatcher stackWatcher = null;
+	private final AEFluidInventory config = new AEFluidInventory( this, 1 );
+
+	public PartFluidLevelEmitter( ItemStack is )
+	{
+		super( is );
+
+		this.getConfigManager().registerSetting( Settings.REDSTONE_EMITTER, RedstoneMode.HIGH_SIGNAL );
+	}
+
+	public long getReportingValue()
+	{
+		return this.reportingValue;
+	}
+
+	public void setReportingValue( final long v )
+	{
+		this.reportingValue = v;
+		this.updateState();
+	}
+
+	@Override
+	public void updateSetting( IConfigManager manager, Enum settingName, Enum newValue )
+	{
+		this.configureWatchers();
+	}
+
+	@Override
+	public void updateWatcher( IStackWatcher newWatcher )
+	{
+		this.stackWatcher = newWatcher;
+		this.configureWatchers();
+	}
+
+	@Override
+	public void onStackChange( IItemList<?> o, IAEStack<?> fullStack, IAEStack<?> diffStack, IActionSource src, IStorageChannel<?> chan )
+	{
+		if( chan == AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) && fullStack.equals( this.config.getFluidInSlot( 0 ) ) )
+		{
+			this.lastReportedValue = fullStack.getStackSize();
+			this.updateState();
+		}
+	}
+
+	@Override
+	public void onFluidInventoryChanged( IAEFluidTank inv, int slot )
+	{
+		this.configureWatchers();
+	}
+
+	@MENetworkEventSubscribe
+	public void channelChanged( final MENetworkChannelsChanged c )
+	{
+		this.updateState();
+	}
+
+	@MENetworkEventSubscribe
+	public void powerChanged( final MENetworkPowerStatusChange c )
+	{
+		this.updateState();
+	}
+
+	@Override
+	public int isProvidingStrongPower()
+	{
+		return this.prevState ? 15 : 0;
+	}
+
+	@Override
+	public int isProvidingWeakPower()
+	{
+		return this.prevState ? 15 : 0;
+	}
+
+	@Override
+	protected int populateFlags( final int cf )
+	{
+		return cf | ( this.prevState ? FLAG_ON : 0 );
+	}
+
+	private void updateState()
+	{
+		final boolean isOn = this.isLevelEmitterOn();
+		if( this.prevState != isOn )
+		{
+			this.getHost().markForUpdate();
+			final TileEntity te = this.getHost().getTile();
+			this.prevState = isOn;
+			Platform.notifyBlocksOfNeighbors( te.getWorld(), te.getPos() );
+			Platform.notifyBlocksOfNeighbors( te.getWorld(), te.getPos().offset( this.getSide().getFacing() ) );
+		}
+	}
+
+	private void configureWatchers()
+	{
+		if( this.stackWatcher != null )
+		{
+			this.stackWatcher.reset();
+
+			final IAEFluidStack myStack = this.config.getFluidInSlot( 0 );
+			if( myStack != null )
+			{
+				this.stackWatcher.add( myStack );
+			}
+		}
+		try
+		{
+			this.updateReportingValue(
+					this.getProxy().getStorage().getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) ) );
+		}
+		catch( GridAccessException e )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)